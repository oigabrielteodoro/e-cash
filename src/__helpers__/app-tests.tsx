import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
