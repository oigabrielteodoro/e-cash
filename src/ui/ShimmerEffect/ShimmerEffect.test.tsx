import React from 'react'
import { render, screen } from '__helpers__/app-tests'

import { ShimmerEffect } from '.'

describe('ShimmerEffect', () => {
  it('should be able match snapshot', () => {
    const { container } = render(<ShimmerEffect isLoading />)

    expect(container).toMatchSnapshot()
  })

  it('should be able render correctly', () => {
    render(<ShimmerEffect isLoading />)

    expect(screen.getByTestId('ShimmerEffect')).toBeInTheDocument()
  })

  it('should be able render element when loading is false', () => {
    render(<ShimmerEffect isLoading={false}>Example</ShimmerEffect>)

    expect(screen.getByText(/example/i)).toBeInTheDocument()
  })
})
