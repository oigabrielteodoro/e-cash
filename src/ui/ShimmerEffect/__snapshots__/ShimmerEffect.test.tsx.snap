// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShimmerEffect should be able match snapshot 1`] = `
.c0 {
  -webkit-animation: ePEqAU 1.2s ease-in-out infinite;
  animation: ePEqAU 1.2s ease-in-out infinite;
  background-image: linear-gradient( 90deg,rgba(0,0,0,0),#fff,rgba(0,0,0,0) );
  background-color: #b3b3b3;
  background-size: 20rem 100%;
  background-repeat: no-repeat;
  min-width: 3.2rem;
  height: 1.6rem;
  opacity: 0.1;
  width: 50%;
  border-radius: 0.5rem;
}

.c0 + span {
  margin-top: 8px;
}

<div>
  <span
    class="c0"
    data-testid="ShimmerEffect"
  />
  <div
    class="Toastify"
  />
</div>
`;
